Vit

A-Jacobson opened the pull request:
ViT-SAM: 74.65 top1 I1k accuracy with 8 devices
ViT: 74.18 top1 I1k accuracy

Questions: @Landanjs I directly imported the model architecture from vit-pytorch. I think it may be a better idea to copy the relevant code as it will be easier to modify in the future to adapt to our methods + easier versioning. Thoughts?

Landanjs commented:
I don't think we should copy the model over, but I don't feel very confident in my answer. Here are my thoughts:
- It may make versioning easier, but we would have to maintain and update the architecture if there are any changes in the original repo (although, I assume there shouldn't be many changes).
- I don't think we should port over a model to make it easier to use our methods since it is likely users will use models from other libraries
- We don't port over models from HuggingFace (should we be using the ViT from HuggingFace?)

@coryMosaicML do you think we should port over the model?

Landanjs reviewed (commented):
Nice straightforward implementation!

Three main points:
1. Do we want to include arguments that can change the model capacity? I think this would avoid additional PRs in the future when we expand to other ViT variations
2. Are the suggested changes to hparams reasonable?
3. Should we postpone merging until adding the additional augmentations to see if the model reaches the reported 74.6% accuracy?

A-Jacobson commented on composer/models/vit_small_patch16/model.py:
```diff
@@ -0,0 +1,40 @@
+        model = ViT(image_size=image_size,
+                    channels=channels,
+                    num_classes=num_classes,
+                    dim=384,  # embed dim/width
+                    patch_size=16,
+                    depth=12,  # layers
+                    heads=6,
+                    mlp_dim=1536,
```
I opted not to do this since this was for a benchmark and I was creating a very specific vit. Changing these parameters would be like changing from a resnet18 -> resnet50 -> resnet101. people generally consider them different models though they're all resnets.

Landanjs replied:
For the ResNets, I would want to move to a single model.py file and several yaml files that vary a depth parameter. This is the setup used for the GPT2 and BERT models. Are there cons to having the models setup this way? The pro is that we can easily vary the model capacity without needing another PR.

A-Jacobson commented on composer/models/vit_small_patch16/model.py:
```diff
@@ -0,0 +1,40 @@
+                 num_classes: int = 1000,
+                 image_size: int = 224,
+                 channels: int = 3,
+                 dropout: Optional[float] = 0.0,
+                 embedding_dropout: Optional[float] = 0.0
```
You are correct! removed.

from https://docs.python.org/3/library/typing.html
On the other hand, if an explicit value of None is allowed, the use of Optional is appropriate, whether the argument is optional or not.

A-Jacobson commented on composer/yamls/models/vit_small_patch16.yaml:
```diff
@@ -0,0 +1,60 @@
+      - 0.9
+      - 0.999
+    eps: 1.0e-08
+    weight_decay: 0.03
```
nice catch, it is in the benchmark run. not sure how this snuck in here.

A-Jacobson commented on composer/models/vit_small_patch16/model.py:
```diff
@@ -0,0 +1,40 @@
+    See this `paper <https://arxiv.org/pdf/2012.12877.pdf>` for details on ViT-S/16.
+
+    Args:
+        num_classes (int): The number of classes for the model.
```
done

A-Jacobson commented:
@Landanjs I'd like to reopen. I updated to 32ep warmup as you suggested at it bumped the accuracy to `74.52` which surpasses the `74.4` reported in the paper (and we're still using less augmentation).

hanlint reviewed (commented):
Do we need to add VIT pytorch to the `setup.py`?

hanlint commented on composer/models/vit_small_patch16/hparams.py:
```diff
@@ -0,0 +1,45 @@
+    embedding_dropout: float = hp.optional("embedding dropout rate", default=0.0)
+
+    def validate(self):
+        if self.image_size is None:
+            raise ValueError("image_size must be specified")
+        if self.channels is None:
+            raise ValueError("channels must be specified")
+        if self.num_classes is None:
+            raise ValueError("num_classes must be specified")
```
Since these are all optional hparams with default,s do we need these here? They will always be set.

A-Jacobson replied:
good point, I'm not sure why I did that. still specify in the docstring that they're optional as well.

A-Jacobson commented:
> Do we need to add VIT pytorch to the `setup.py`?

I just added it. I had previously added it + timm to composer[vision] when I made this pr but it seems that the requirement was lost when we refactored imports.

Landanjs commented on composer/models/vit_small_patch16/hparams.py:
```diff
@@ -0,0 +1,45 @@
+    num_classes: int = hp.optional("number of classes.  Needed for classification tasks", default=1000)
+    image_size: int = hp.optional(
+        "input image size. If you have rectangular images, make sure your image size is the maximum of the width and height",
+        default=244)
```
Should be 224

Landanjs reviewed (approved):
LGTM! Should make one default change
